File: ios/ContractRunner/Sources/ContractRunner/OmdbNormalizer.swift
import Foundation

public struct OmdbRatingInput: Equatable {
    public let source: String?
    public let value: String?

    public init(source: String?, value: String?) {
        self.source = source
        self.value = value
    }
}

public struct OmdbRating: Equatable {
    public let source: String
    public let value: String

    public init(source: String, value: String) {
        self.source = source
        self.value = value
    }
}

public struct OmdbDetails: Equatable {
    public let ratings: [OmdbRating]
    public let metascore: String?
    public let imdbRating: String?
    public let imdbVotes: String?
    public let type: String?

    public init(
        ratings: [OmdbRating] = [],
        metascore: String? = nil,
        imdbRating: String? = nil,
        imdbVotes: String? = nil,
        type: String? = nil
    ) {
        self.ratings = ratings
        self.metascore = metascore
        self.imdbRating = imdbRating
        self.imdbVotes = imdbVotes
        self.type = type
    }
}

public func normalizeOmdbImdbId(_ value: String?) -> String? {
    let normalized = value?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
    guard normalized.range(of: "^tt\\d+$", options: [.regularExpression, .caseInsensitive]) != nil else {
        return nil
    }
    return normalized
}

public func normalizeOmdbDetails(
    ratings: [OmdbRatingInput],
    metascore: String?,
    imdbRating: String?,
    imdbVotes: String?,
    type: String?
) -> OmdbDetails {
    let normalizedMetascore = normalizedOmdbField(metascore)
    let normalizedImdbRating = normalizedOmdbField(imdbRating)
    let normalizedImdbVotes = normalizedOmdbField(imdbVotes)
    let normalizedType = normalizedOmdbField(type)

    var dedupedRatings: [OmdbRating] = []
    var seenSources = Set<String>()

    for rating in ratings {
        guard let source = normalizedOmdbField(rating.source),
              let value = normalizedOmdbField(rating.value) else {
            continue
        }
        let key = source.lowercased()
        guard seenSources.insert(key).inserted else {
            continue
        }
        dedupedRatings.append(OmdbRating(source: source, value: value))
    }

    let imdbKey = internetMovieDatabaseSource.lowercased()
    if let normalizedImdbRating, seenSources.insert(imdbKey).inserted {
        dedupedRatings.append(
            OmdbRating(source: internetMovieDatabaseSource, value: "\(normalizedImdbRating)/10")
        )
    }

    let metacriticKey = metacriticSource.lowercased()
    if let normalizedMetascore, seenSources.insert(metacriticKey).inserted {
        dedupedRatings.append(
            OmdbRating(source: metacriticSource, value: "\(normalizedMetascore)/100")
        )
    }

    return OmdbDetails(
        ratings: dedupedRatings,
        metascore: normalizedMetascore,
        imdbRating: normalizedImdbRating,
        imdbVotes: normalizedImdbVotes,
        type: normalizedType
    )
}

private func normalizedOmdbField(_ value: String?) -> String? {
    guard let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines),
          !trimmed.isEmpty,
          trimmed.caseInsensitiveCompare("N/A") != .orderedSame else {
        return nil
    }
    return trimmed
}

private let internetMovieDatabaseSource = "Internet Movie Database"
private let metacriticSource = "Metacritic"

File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/metadata/OmdbNormalizer.kt
package com.crispy.tv.domain.metadata

import java.util.Locale

data class OmdbRatingInput(
    val source: String?,
    val value: String?,
)

data class OmdbRating(
    val source: String,
    val value: String,
)

data class OmdbDetails(
    val ratings: List<OmdbRating> = emptyList(),
    val metascore: String? = null,
    val imdbRating: String? = null,
    val imdbVotes: String? = null,
    val type: String? = null,
)

fun normalizeOmdbImdbId(value: String?): String? {
    val normalized = value?.trim()?.lowercase(Locale.US).orEmpty()
    return normalized.takeIf { IMDB_ID_REGEX.matches(it) }
}

fun normalizeOmdbDetails(
    ratings: List<OmdbRatingInput>,
    metascore: String?,
    imdbRating: String?,
    imdbVotes: String?,
    type: String?,
): OmdbDetails {
    val normalizedMetascore = metascore.normalizedOmdbField()
    val normalizedImdbRating = imdbRating.normalizedOmdbField()
    val normalizedImdbVotes = imdbVotes.normalizedOmdbField()
    val normalizedType = type.normalizedOmdbField()

    val dedupedRatings = mutableListOf<OmdbRating>()
    val seenSources = linkedSetOf<String>()

    ratings.forEach { rating ->
        val source = rating.source.normalizedOmdbField() ?: return@forEach
        val value = rating.value.normalizedOmdbField() ?: return@forEach
        val normalizedSource = source.lowercase(Locale.US)
        if (!seenSources.add(normalizedSource)) {
            return@forEach
        }
        dedupedRatings += OmdbRating(source = source, value = value)
    }

    if (normalizedImdbRating != null && seenSources.add(INTERNET_MOVIE_DATABASE_SOURCE.lowercase(Locale.US))) {
        dedupedRatings += OmdbRating(
            source = INTERNET_MOVIE_DATABASE_SOURCE,
            value = "$normalizedImdbRating/10",
        )
    }

    if (normalizedMetascore != null && seenSources.add(METACRITIC_SOURCE.lowercase(Locale.US))) {
        dedupedRatings += OmdbRating(
            source = METACRITIC_SOURCE,
            value = "$normalizedMetascore/100",
        )
    }

    return OmdbDetails(
        ratings = dedupedRatings,
        metascore = normalizedMetascore,
        imdbRating = normalizedImdbRating,
        imdbVotes = normalizedImdbVotes,
        type = normalizedType,
    )
}

private fun String?.normalizedOmdbField(): String? {
    val normalized = this?.trim().orEmpty()
    if (normalized.isBlank() || normalized.equals("N/A", ignoreCase = true)) {
        return null
    }
    return normalized
}

private const val INTERNET_MOVIE_DATABASE_SOURCE = "Internet Movie Database"
private const val METACRITIC_SOURCE = "Metacritic"
private val IMDB_ID_REGEX = Regex("^tt\\d+$", RegexOption.IGNORE_CASE)
